### ISSUE TITLE
Angular Language Service ngcc error in VSCode 1.62.1 and above

### ISSUE THREAD

**Issue by jkyoutsey** [angular/angular#44148]
### Which @angular/* package(s) are the source of the bug?

language-service

### Is this a regression?

Yes

### Description

I've upgraded to Angular 12 and I've installed the latest 12.x version VSIX of @angular/language-service extension in VSCode. But I get a popup error in the bottom right corner.

```
Angular extension might not work correctly because ngcc operation failed. Try to invoke ngcc manually by running 'npm/yarn run ngcc'. Please see the extension output for more information.
```

### Please provide the exception or error you saw

```true
This is the OUTPUT:
[Info  - 7:04:14 AM] Angular language server process ID: 34020
[Info  - 7:04:14 AM] Using typescript/lib/tsserverlibrary v4.3.5 from /Users/jyoutsey/src/myorg/ng/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 7:04:14 AM] Using @angular/language-service v12.2.12 from /Users/jyoutsey/.vscode/extensions/angular.ng-template-12.2.3/node_modules/@angular/language-service/index.js
[Info  - 7:04:14 AM] Logging is turned off. To enable, run command 'Open Angular server log'.
[Error - 7:04:39 AM] Failed to run ngcc for /Users/jyoutsey/src/myorg/ng/tsconfig.json, language service may not operate correctly:
    ngcc for /Users/jyoutsey/src/myorg/ng/tsconfig.json returned exit code null, stderr: 
[Info  - 7:04:39 AM] Enabling Ivy language service for /Users/jyoutsey/src/myorg/ng/tsconfig.json.
```

### Please provide the environment you discovered this bug in

```true
Angular CLI: 12.2.13
Node: 14.15.1
Package Manager: npm 6.14.11
OS: darwin x64

Angular: 12.2.13
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.1202.13
@angular-devkit/build-angular      12.2.13
@angular-devkit/core               12.2.13
@angular-devkit/schematics         12.2.13
@angular/cdk                       12.2.12
@angular/material                  12.2.12
@angular/material-moment-adapter   12.2.12
@schematics/angular                12.2.13
rxjs                               7.4.0
typescript                         4.3.5
```

### Anything else?

I've uninstalled the extension, restarted VSCode, and installed it again, to no avail.

**Comment by atscott** [comment]
A few questions to help debugging, as I have not been able to reproduce this myself:

* Do you have an example project and what are the steps to reproduce with an example project?
* Are your v12 projects using Ivy or View Engine? (is `"enableIvy": false` in the `tsconfig.json`)
* Are the dependencies installed for the project? (`yarn/npm install`)
* Does `ngc` fail when run manually on the project? (`yarn/npm run ngc` or `./node_modules/.bin/ngc tsconfig.json`)
* Does `ngcc` fail when run manually on the project? (`yarn/npm run ngcc` or `./node_modules/.bin/ngcc --tsconfig tsconfig.json`)

**Comment by jkyoutsey** [comment]
@atscott Inline...

- Do you have an example project and what are the steps to reproduce with an example project?
  - No, this is a very large project.
- Are your v12 projects using Ivy or View Engine? (is "enableIvy": false in the tsconfig.json)
  - Ivy. No entry in tsconfig.json
- Are the dependencies installed for the project? (yarn/npm install)
  - Yes
- Does ngc fail when run manually on the project? (yarn/npm run ngc or ./node_modules/.bin/ngc tsconfig.json)
  - No. Only spec files fail and only where we have a TestComponent defined because it can't seem to resolve the references. However `ng test` and `ng test --configuration production` and WallabyJS all run tests without issues. `npm install` runs nccc without issue as well. And everything about the application works.
- Does ngcc fail when run manually on the project? (yarn/npm run ngcc or ./node_modules/.bin/ngcc --tsconfig tsconfig.json)
  - `./node_modules/.bin/ngcc --tsconfig tsconfig.json` produces no output at all when I run it.
  
Here is my tsconfig.json
```
{
	"compileOnSave": false,
	"compilerOptions": {
		"baseUrl": "./",
		"importHelpers": true,
		"outDir": "./dist/out-tsc",
		"sourceMap": true,
		"declaration": false,
		"module": "es2020",
		"moduleResolution": "node",
		"emitDecoratorMetadata": true,
		"experimentalDecorators": true,
		"target": "es2015",
		"typeRoots": ["node_modules/@types"],
		"lib": ["es2017", "dom"],
		"paths": {
			"@environments/*": ["src/environments/*"],
			"@services/*": ["src/app/services/*"],
			"@core/*": ["src/app/core/*"],
			"@utils/*": ["src/app/utils/*"],
			"@common/*": ["src/app/common/*"],
			"@store/*": ["src/app/store/*"],
			"@account/*": ["src/app/features/account/*"],
			"@chrome/*": ["src/app/features/chrome/*"],
			"@media/*": ["src/app/features/media/*"],
			"@book/*": ["src/app/features/book/*"]
		},
		"esModuleInterop": true,
		"allowSyntheticDefaultImports": true,

		// Strictness Checks
		"noImplicitThis": true,
		"alwaysStrict": true
	},
	"angularCompilerOptions": {
		"fullTemplateTypeCheck": true
	}
}
```

**Comment by atscott** [comment]
@jkyoutsey I don't see anything abnormal about any of the information you've provided. You mentioned that this is a regression and was working before. Let's see if we can pin down what version this stopped working with. Can you try installing different versions of angular / the vscode extension to see where the problem started occurring?

**Comment by jkyoutsey** [comment]
@atscott It was working in 11 latest. I upgraded to 12 latest via CLI. At one point I had installed v13 of the extension, then got that error. Removed the extension and installed v12.2.12 of the extension via VSIX. 
I'm not sure how I would go backwards from Angular 12 to 11. But I could install VSIX versions of the extension going back to 12.0?

**Comment by atscott** [comment]
@jkyoutsey Right, if you can try going backwards with the vscode extension versions, maybe we can find one that worked. If not, then it must have been the update from v11 to v12 of Angular. You could manually change the versions in the `package.json` with find/replace to go backwards. It might be that there was a specific patch version (`12.2.x`) that broke things.

**Comment by atscott** [comment]
@jkyoutsey What version of VSCode are you using? I'm also gathering information from the report on the extension repository and it seems that it might be related to the latest Insider version. Can you try the extension an older version of vscode as well?

**Comment by atscott** [comment]
Okay, I have been able to isolate this to the vscode `1.62.1` release. The error does not occur in the `1.62.0` release. When updating to `1.62.1` with no other changes, the `ngcc` process fails.

**Comment by jkyoutsey** [comment]
@atscott Thanks for figuring out the source of this!

**Comment by geokaps** [comment]
I also have this issue after upgrading to vscode 1.62.1, Angular Language Service v13.0.0, using Nx repos just like others above, etc.

However, this is only happening on VSCode 1.62.1 on MacOS (Big Sur) and not on my Windows 10 box/WSL instance. Same exact Nx Repo, on same branch, same extensions, same VSCode 1.62.1 but for Windows. Not sure if this is helpful or not, but **it is only occurring on my Mac.**  

Both machines are configured identically:

Windows WSL Ubuntu:
```
Angular CLI: 12.2.6
Node: 14.17.4
Package Manager: npm 7.22.0
OS: linux x64

Angular: 12.2.6
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1202.6
@angular-devkit/build-angular   12.2.6
@angular-devkit/core            12.2.6
@angular-devkit/schematics      12.2.6
@schematics/angular             12.2.6
rxjs                            6.6.7
typescript                      4.3.5
```

 MacOS Big Sur:
```
Angular CLI: 12.2.6
Node: 14.17.4
Package Manager: npm 7.22.0
OS: darwin x64

Angular: 12.2.6
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1202.6
@angular-devkit/build-angular   12.2.6
@angular-devkit/core            12.2.6
@angular-devkit/schematics      12.2.6
@schematics/angular             12.2.6
rxjs                            6.6.7
typescript                      4.3.5
```



Output on Mac:
```
[Info  - 5:54:11 PM] Angular language server process ID: 24412
[Info  - 5:54:11 PM] Using typescript/lib/tsserverlibrary v4.3.5 from /Users/georgekapsambelis/Repos/esg-project/UI/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 5:54:11 PM] Using @angular/language-service v13.0.0 from /Users/georgekapsambelis/.vscode/extensions/angular.ng-template-13.0.0/node_modules/@angular/language-service/index.js
[Info  - 5:54:11 PM] Logging is turned off. To enable, run command 'Open Angular server log'.
[Error - 5:54:47 PM] Failed to run ngcc for /Users/georgekapsambelis/Repos/esg-project/UI/libs/credit-alert-app/feature-credit-alert/tsconfig.lib.json, language service may not operate correctly:
    ngcc for /Users/georgekapsambelis/Repos/esg-project/UI/libs/credit-alert-app/feature-credit-alert/tsconfig.lib.json returned exit code null, stderr: 
[Info  - 5:54:47 PM] Enabling Ivy language service for /Users/georgekapsambelis/Repos/esg-project/UI/libs/credit-alert-app/feature-credit-alert/tsconfig.lib.json.
```

Output on Windows:
```
[Info  - 6:03:43 PM] Angular language server process ID: 44573
[Info  - 6:03:43 PM] Using typescript/lib/tsserverlibrary v4.3.5 from /home/gkapsambelis/Repos/esg-project/UI/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 6:03:43 PM] Using @angular/language-service v13.0.0 from /home/gkapsambelis/.vscode-server/extensions/angular.ng-template-13.0.0/node_modules/@angular/language-service/index.js
[Info  - 6:03:43 PM] Logging is turned off. To enable, run command 'Open Angular server log'.
[Info  - 6:03:47 PM] Enabling Ivy language service for /home/gkapsambelis/Repos/esg-project/UI/libs/credit-alert-app/feature-credit-alert/tsconfig.lib.json.
```

**Look at the output for the Mac** - strange that ngcc fails and then right after is does a "Enabling Ivy language service" on same file.

**Comment by atscott** [comment]
> Look at the output for the Mac - strange that ngcc fails and then right after is does a "Enabling Ivy language service" on same file.

Right, we still enable the language service even if ngcc fails because you might be able to resolve the issue yourself. The popup error mentions "Try to invoke ngcc manually by running 'npm/yarn run ngcc'". If that works, then the language service can still function correctly. In reality, `ngcc` doesn't _have_ to run all the time. If you're running your application, it likely already ran `ngcc` so the language service doesn't necessarily need to. We just can't rely on this being the case.

**Comment by ColonelPanic-42** [comment]
Can confirm this is an issue on Manjaro Linux as well, in both 1.62.1 and 1.62.2:

```
[Info  - 10:17:58 am] Angular language server process ID: 577614
[Info  - 10:17:58 am] Using typescript/lib/tsserverlibrary v4.3.4 from /home/user/.vscode/extensions/angular.ng-template-12.2.3/node_modules/typescript/lib/tsserverlibrary.js
[Info  - 10:17:58 am] Using @angular/language-service v12.2.12 from /home/user/.vscode/extensions/angular.ng-template-12.2.3/node_modules/@angular/language-service/index.js
[Info  - 10:17:58 am] Logging is turned off. To enable, run command 'Open Angular server log'.
[Error - 10:18:08 am] Failed to run ngcc for /home/user/project/@app/client/tsconfig.json, language service may not operate correctly:
    ngcc for /home/user/project/@app/client/tsconfig.json returned exit code null, stderr: 
[Info  - 10:18:08 am] Enabling Ivy language service for /home/user/project/@app/client/tsconfig.json.
```

```
Angular CLI: 12.2.12
Node: 14.17.5
Package Manager: npm 6.14.14
OS: linux x64

Angular: 12.2.12
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router, service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1202.12
@angular-devkit/build-angular   12.2.12
@angular-devkit/core            12.2.12
@angular-devkit/schematics      12.2.12
@schematics/angular             12.2.12
ng-packagr                      12.2.5
rxjs                            6.5.5
typescript                      4.3.5
```

Related issues:
https://github.com/angular/vscode-ng-language-service/issues/1565
https://github.com/microsoft/vscode/issues/136872 (closed, not their problem apparently even though it was their update that triggered it)

**Comment by geokaps** [comment]
Just wanted to add as others have noted, rolling back to the September 2021 v1.61.2 VSCode release fixes this issue for me on the Mac.  https://code.visualstudio.com/updates/v1_61

**Comment by atscott** [comment]
Downgrading the severity here to P3. We are still actively investigating this issue and are tracking https://github.com/microsoft/vscode/issues/136988 as the likely resolution. 

In the meantime, the error we show does give the appropriate information. `ngcc` failed to run, which _might_ cause the language service to not work correctly but can be resolved by `npm/yarn run ngcc`. Or if you are already running a dev server, `ngcc` likely already ran and you don't have to do anything.

**Comment by RobertAKARobin** [comment]
Looks like VSCode has released a fix. Does that unblock this issue?

**Comment by atscott** [comment]
@RobertAKARobin  The fix was included in their latest insider version release but I've been leaving this open for visibility until their recovery release this week (1.62.3 I believe). There's nothing to be done on our end since it was a bug in VSCode.

**Comment by atscott** [comment]
Closing as resolved in the `1.62.3` release of VSCode.
